pkg/controllermanager/controller/credentialsbinding/reconciler.go

```diff
@@ -0,0 +1,258 @@
+						if err := controllerutils.RemoveFinalizers(ctx, r.Client, secret, gardencorev1beta1.ExternalGardenerName); err != nil {
+							return reconcile.Result{}, fmt.Errorf("failed to remove finalizer from secret: %w", err)
+						}
+					}
+				} else if !apierrors.IsNotFound(err) {
+					return reconcile.Result{}, err
+				}
+			}
+
+			if err := r.removeLabelFromQuotas(ctx, credentialsBinding); err != nil {
+				return reconcile.Result{}, err
+			}
+
+			// Remove finalizer from CredentialsBinding
+			if controllerutil.ContainsFinalizer(credentialsBinding, gardencorev1beta1.GardenerName) {
+				log.Info("Removing finalizer")
+				if err := controllerutils.RemoveFinalizers(ctx, r.Client, credentialsBinding, gardencorev1beta1.GardenerName); err != nil {
+					return reconcile.Result{}, fmt.Errorf("failed to remove finalizer: %w", err)
+				}
+			}
+
+			return reconcile.Result{}, nil
+		}
+
+		message := fmt.Sprintf("Cannot delete CredentialsBinding, because the following Shoots are still referencing it: %+v", associatedShoots)
+		r.Recorder.Event(credentialsBinding, corev1.EventTypeNormal, v1beta1constants.EventResourceReferenced, message)
+		return reconcile.Result{}, errors.New(message)
+	}
+
+	if !controllerutil.ContainsFinalizer(credentialsBinding, gardencorev1beta1.GardenerName) {
+		log.Info("Adding finalizer")
+		if err := controllerutils.AddFinalizers(ctx, r.Client, credentialsBinding, gardencorev1beta1.GardenerName); err != nil {
+			return reconcile.Result{}, fmt.Errorf("could not add finalizer: %w", err)
+		}
+	}
+
+	// TODO(dimityrmirchev): this code should eventually handle workload identities as a valid credential ref
+	// Add the Gardener finalizer to the referenced Secret
+	// to protect it from deletion as long as the CredentialsBinding resource exists.
+	secret := &corev1.Secret{}
+	if err := r.Client.Get(ctx, client.ObjectKey{Namespace: credentialsBinding.CredentialsRef.Namespace, Name: credentialsBinding.CredentialsRef.Name}, secret); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if !controllerutil.ContainsFinalizer(secret, gardencorev1beta1.ExternalGardenerName) {
+		log.Info("Adding finalizer to secret", "secret", client.ObjectKeyFromObject(secret))
+		if err := controllerutils.AddFinalizers(ctx, r.Client, secret, gardencorev1beta1.ExternalGardenerName); err != nil {
+			return reconcile.Result{}, fmt.Errorf("could not add finalizer to secret: %w", err)
+		}
+	}
+
+	labelKey := v1beta1constants.LabelShootProviderPrefix + credentialsBinding.Provider.Type
+	if !metav1.HasLabel(secret.ObjectMeta, labelKey) {
+		patch := client.MergeFrom(secret.DeepCopy())
+		metav1.SetMetaDataLabel(&secret.ObjectMeta, labelKey, "true")
+		if err := r.Client.Patch(ctx, secret, patch); err != nil {
+			return reconcile.Result{}, fmt.Errorf("failed to add provider type label to Secret referenced in CredentialsBinding: %w", err)
+		}
+	}
+
+	if len(credentialsBinding.Quotas) != 0 {
```

Comment:
if the slice is empty, then the for loop body will not be executed even once, I mean this check is obsolete. 

```suggestion
```